inject/src/main/java/io/micronaut/inject/ast/ClassElement.java

```diff
@@ -172,8 +209,117 @@ default boolean isProxy() {
+    default Optional<MethodElement> getDefaultConstructor() {
+        Optional<MethodElement> staticCreator = findDefaultStaticCreator();
+        if (staticCreator.isPresent()) {
+            return staticCreator;
+        }
+        if (isInner() && !isStatic()) {
+            // only static inner classes can be constructed
+            return Optional.empty();
+        }
+        List<ConstructorElement> constructors = findConstructors()
+            .stream()
+            .filter(ctor -> ctor.getParameters().length == 0)
+            .collect(Collectors.toList());
+        if (constructors.isEmpty()) {
+            return Optional.empty();
+        }
+        if (constructors.size() == 1) {
+            return Optional.of(constructors.get(0));
+        }
+        return constructors.stream()
+            .filter(Element::isPublic)
+            .<MethodElement>map(c -> c)
+            .findFirst();
+
+    }
+
+    /**
+     * Find static creator.
+     *
+     * @return a static creator
+     * @since 4.0.0
+     */
+    default Optional<MethodElement> findStaticCreator() {
+        List<MethodElement> staticCreators = findStaticCreators();
+        if (staticCreators.isEmpty()) {
+            return Optional.empty();
+        }
+        if (staticCreators.size() == 1) {
+            return Optional.of(staticCreators.get(0));
+        }
+        //Can be multiple static @Creator methods. Prefer one with args here. The no arg method (if present) will
+        //be picked up by staticDefaultCreatorFor
+        List<MethodElement> withArgs = staticCreators.stream().filter(method -> method.getParameters().length > 0).collect(Collectors.toList());
+        if (withArgs.size() == 1) {
+            return Optional.of(withArgs.get(0));
+        } else {
+            staticCreators = withArgs;
+        }
+        return staticCreators.stream().filter(Element::isPublic).findFirst();
+    }
+
+    /**
+     * Find default static creator.
+     *
+     * @return a static creator
+     * @since 4.0.0
+     */
+    default Optional<MethodElement> findDefaultStaticCreator() {
+        List<MethodElement> staticCreators = findStaticCreators()
+            .stream()
+            .filter(c -> c.getParameters().length == 0)
+            .collect(Collectors.toList());
+        if (staticCreators.isEmpty()) {
+            return Optional.empty();
+        }
+        if (staticCreators.size() == 1) {
+            return Optional.of(staticCreators.get(0));
+        }
+        return staticCreators.stream().filter(Element::isPublic).findFirst();
+    }
+
+    /**
+     * Find accessible constructors.
+     *
+     * @return accessible constructors
+     * @since 4.0.0
+     */
+    default List<ConstructorElement> findConstructors() {
+        return getEnclosedElements(ElementQuery.CONSTRUCTORS)
+            .stream()
+            .filter(ctor -> !ctor.isPrivate())
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Find static creators.
```

Comment:
clarify in the javadoc what a `static` creator is and add references to the other related methods